chainstate/src/detail/tokens/mod.rs

```diff
@@ -0,0 +1,133 @@
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::{
+    chain::{
+        config::{TOKEN_MAX_DEC_COUNT, TOKEN_MAX_TICKER_LEN, TOKEN_MAX_URI_LEN},
+        tokens::{TokenData, TokensError},
+        Block, Transaction,
+    },
+    primitives::{Amount, Id, Idable},
+};
+use utils::ensure;
+
+pub fn check_tokens_transfer_data(
+    source_block_id: Id<Block>,
+    tx: &Transaction,
+    amount: &Amount,
+) -> Result<(), TokensError> {
+    // Check amount
+    ensure!(
+        amount > &Amount::from_atoms(0),
+        TokensError::TransferZeroTokens(tx.get_id(), source_block_id)
+    );
+
+    Ok(())
+}
+
+pub fn check_tokens_burn_data(
+    tx: &Transaction,
+    source_block_id: &Id<Block>,
+    amount_to_burn: &Amount,
+) -> Result<(), TokensError> {
+    // Check amount
+    ensure!(
+        amount_to_burn != &Amount::from_atoms(0),
+        TokensError::BurnZeroTokens(tx.get_id(), *source_block_id)
+    );
+    Ok(())
+}
+
+pub fn check_tokens_issuance_data(
+    token_ticker: &Vec<u8>,
+    amount_to_issue: &Amount,
+    number_of_decimals: &u8,
+    metadata_uri: &Vec<u8>,
+    tx_id: Id<Transaction>,
+    source_block_id: Id<Block>,
+) -> Result<(), TokensError> {
+    // Check token name
+    if token_ticker.len() > TOKEN_MAX_TICKER_LEN
+        || token_ticker.is_empty()
+        || !String::from_utf8_lossy(token_ticker).is_ascii()
+    {
+        return Err(TokensError::IssueErrorIncorrectTicker(
+            tx_id,
+            source_block_id,
+        ));
+    }
+
+    // Check amount
+    if amount_to_issue == &Amount::from_atoms(0) {
+        return Err(TokensError::IssueErrorIncorrectAmount(
+            tx_id,
+            source_block_id,
+        ));
+    }
+
+    // Check decimals
+    if number_of_decimals > &TOKEN_MAX_DEC_COUNT {
+        return Err(TokensError::IssueErrorTooManyDecimals(
+            tx_id,
+            source_block_id,
+        ));
+    }
+
+    // Check URI
+    if metadata_uri.len() > TOKEN_MAX_URI_LEN || !String::from_utf8_lossy(metadata_uri).is_ascii() {
+        return Err(TokensError::IssueErrorIncorrectMetadataURI(
+            tx_id,
+            source_block_id,
+        ));
+    }
+    Ok(())
+}
+
+pub fn check_tokens_data(
+    token_data: &TokenData,
+    tx: &Transaction,
+    source_block_id: Id<Block>,
+) -> Result<(), TokensError> {
+    match token_data {
+        TokenData::TokenTransferV1 {
+            token_id: _,
+            amount,
+        } => {
+            check_tokens_transfer_data(source_block_id, tx, amount)?;
+        }
+        TokenData::TokenIssuanceV1 {
+            token_ticker,
```

Comment:
Please also add a public key, basically the controller public key. The controller will have rights to issue more tokens later. It can be an `Option<PublicKey>` because issuers may decide to not allow that.

Reply:
You know what, don't do that now. Let's postpone this to next PR. Create an issue for it.

In that issue, please reference this comment. I'll add more information in another comment.

Reply:
Keep in mind that the total supply should be taken from another place if this is enabled. For example, some storage in database. This should also support reorgs and such complex operations with newly minted coins after initial issuance being subtracted when disconnecting a block.

Reply:
Done